Hw04 lru cache

zamatay opened the pull request.

Commit:
init

Commit:
repair bug!

Commit:
repair bug!

zamatay committed:
task 04

hungry-honeybadger commented on hw04_lru_cache/cache.go:
```diff
@@ -1,25 +1,86 @@
+	value, isOk := r.items[key]
+	r.mu.Unlock()
+	if isOk {
+		r.queue.MoveToFront(value)
```
Операция MoveToFront не является потокобезопасной

hungry-honeybadger commented on hw04_lru_cache/cache.go:
```diff
@@ -1,25 +1,86 @@
+
+func (r *lruCache) Clear() {
+	r.mu.Lock()
+	r.capacity = 0
```
Присваивать капасити 0 делает кэш бесполезным, а не очищает кэш. Предлагаю это строчку убрать

hungry-honeybadger commented on hw04_lru_cache/cache.go:
```diff
@@ -1,25 +1,86 @@
+}
+
+func (r *lruCache) remove(item *ListItem) {
+	v := r.queue.Back()
```
В данном случае `item==v`. Зачем хранить 2 раза 1 и тоже значение?

Commit:
fix

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw04_lru_cache
Your branch is up to date with 'origin/hw04_lru_cache'.

Changes to be committed:
	modified:   hw04_lru_cache/list.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw04_lru_cache
Your branch is up to date with 'origin/hw04_lru_cache'.

Changes to be committed:
	modified:   hw04_lru_cache/cache.go

hungry-honeybadger commented on hw04_lru_cache/list.go:
```diff
@@ -1,26 +1,106 @@
+}
+
+type list struct {
+	mu        sync.Mutex
```
По условиям задания потокобезопасным нужно сделать кэш .Двусвязный список модифицировать нельзя.

hungry-honeybadger reviewed (changes requested):
Отличная работа. Но есть пара замечаний. Пока работу принять не могу.

Commit:
fix

hungry-honeybadger commented on hw04_lru_cache/cache.go:
```diff
@@ -1,25 +1,84 @@
+
+func (r *lruCache) Clear() {
+	r.mu.Lock()
+	r.items = make(map[Key]*ListItem, 0)
```
В данном случае мы заранее знаем максимальный размера мапы и можем его сразу задать. Это позволит избежать лишних операций с памятью

hungry-honeybadger reviewed (approved):
Отлично. Работа принята. 10/10

zamatay committed:
Merge branch 'master' into hw04_lru_cache

Commit:
repair bug!

Commit:
Merge remote-tracking branch 'origin/hw04_lru_cache' into hw04_lru_cache
